v21.2/connection-reference.md

```diff
@@ -0,0 +1,732 @@
+---
+title: Client Connection Reference
+summary: How to connect to a CockroachDB cluster from different tools
+toc: true
+---
+
+<div class="filters clearfix">
+  <button class="filter-button page-level" data-scope="cockroachcloud">Use {{ site.data.products.serverless-plan }}</button>
+  <button class="filter-button page-level" data-scope="local">Use a Local Cluster</button>
+</div>
+
+This page documents the required connection configuration for each [fully-supported, third-party tool](third-party-database-tools.html).
+
+For a list of all supported cluster connection parameters, see the [`cockroach` Connection Parameters](connection-parameters.html).
+
+<div class="filters clearfix">
+  <button class="filter-button page-level" data-scope="python">Python</button>
+  <button class="filter-button page-level" data-scope="java">Java</button>
+  <button class="filter-button page-level" data-scope="go">Go</button>
+  <button class="filter-button page-level" data-scope="ruby">Ruby</button>
+  <button class="filter-button page-level" data-scope="js-ts">JavaScript/TypeScript</button>
+</div>
+
+<section class="filter-content" markdown="1" data-scope="python">
+
+<div class="filters clearfix">
+  <button class="filter-button page-level" data-scope="psycopg">Psycopg</button>
+  <button class="filter-button page-level" data-scope="sqlalchemy">SQLAlchemy</button>
+  <button class="filter-button page-level" data-scope="django">Django</button>
+</div>
+
+{{site.data.alerts.callout_danger}}
+To connect to a {{ site.data.products.serverless }} cluster from a Python application, you must have a valid CA certificate located at `~/.postgresql/root.crt`.
+
+For instructions on downloading a CA certificate from the {{ site.data.products.db }} Console, see [Connect to a {{ site.data.products.serverless }} Cluster](../cockroachcloud/connect-to-a-serverless-cluster.html).
+{{site.data.alerts.end}}
+
+<section class="filter-content" markdown="1" data-scope="psycopg">
+
+## Connect with Psycopg
+
+To connect to CockroachDB with [Psycopg](https://www.psycopg.org), pass a connection string to the [`psycopg.connect` function](https://www.psycopg.org/psycopg3/docs/api/module.html#psycopg.connect).
+
+For example:
+
+{% include copy-clipboard.html %}
+~~~ python
+import psycopg
+
+conn = psycopg.connect('{connection-string}')
+~~~
+
+Psycopg accepts the following format for CockroachDB connection strings:
```

Comment:
What I was quite unsuccessfully attempting to articulate here is that, in an ideal world, we wouldn't even need to document the format of CockroachDB connection strings for Serverless because the user would simply copy the appropriate connection string directly from the CockroachDB Cloud web UI. However, I am now realizing that this is not immediately possible because the web UI doesn't generate the correct connection strings. So, I guess the first step would be for the web UI to generate the correct connection string for a particular tool. Once that's ready to go, then we could un-document this.